tests/datasets/test_streaming.py

```diff
@@ -0,0 +1,163 @@
+import math
+import os
+import shutil
+import tempfile
+import time
+from typing import Dict, Iterable, List, Optional
+
+import numpy as np
+import pytest
+import torch
+from torch.utils.data import DataLoader
+
+from composer.datasets.streaming import StreamingDataset, StreamingDatasetWriter
+
+
+def get_fake_samples_decoders(num_samples: int) -> List[Dict[str, bytes]]:
+    samples = [{"uid": f"{ix:06}".encode("utf-8"), "data": (3 * ix).to_bytes(4, "big")} for ix in range(num_samples)]
+    decoders = {
+        "uid": lambda uid_bytes: uid_bytes.decode("utf-8"),
+        "data": lambda data_bytes: int.from_bytes(data_bytes, "big")
+    }
+    return samples, decoders
+
+
+def write_synthetic_streaming_dataset(samples: List[Dict[str, bytes]], shard_size_limit: int) -> str:
+    tmpdir = tempfile.TemporaryDirectory().name
+    first_sample_fields = list(samples[0].keys())
+    with StreamingDatasetWriter(dirname=tmpdir, fields=first_sample_fields,
+                                shard_size_limit=shard_size_limit) as writer:
+        writer.write_samples(samples=samples)
+    return tmpdir
+
+
+@pytest.mark.parametrize("num_samples", [100, 10000])
+@pytest.mark.parametrize("shard_size_limit", [1 << 8, 1 << 16, 1 << 24])
```

Comment:
Curious what value these parameterization combinations add? Would it make sense to specify the (num_samples, shard_size_limit) combinations together to ensure that all the edge cases are covered? Not a huge fan of increasing the test size unless if there is specific value that comes out of it...

Reply:
These were just a few random values, I think in practice this test takes <1s? Unfortunately I don't really know how to construct good edge cases without perfetctly controlling the bytes per shard, which is pretty hard to do given headers and all... can we leave as is?

Reply:
I see. So if the test is to ensure that shards are flushed properly, do we need the parameterization matrix? Or can we accomplish this through one test?

Reply:
I think the matrix is to make sure that samples are flushed and the correct number of shard are written, which means we want to check a few cases of samples % shard_bytes.